runtime/gc_include/ObjectAccessBarrierAPI.hpp

```diff
@@ -341,25 +342,21 @@ class MM_ObjectAccessBarrierAPI
 	}
 
 	VMINLINE void
-	cloneArray(J9VMThread *currentThread, j9object_t original, j9object_t copy, J9Class *objectClass, U_32 size)
+	cloneArray(J9VMThread *currentThread, j9object_t original, j9object_t copy, J9Class *objectClass, U_32 size, MM_objectMapFunction objectMapFunction = NULL, void *objectMapData = NULL, bool initializeLockWord = true)
 	{
+		/* Note: initializeLockWord is ignored as arrays never have inline lockwords */
 #if defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER) 
-		currentThread->javaVM->memoryManagerFunctions->j9gc_objaccess_cloneIndexableObject(currentThread, (J9IndexableObject*)original, (J9IndexableObject*)copy);
+		currentThread->javaVM->memoryManagerFunctions->j9gc_objaccess_cloneIndexableObject(currentThread, (J9IndexableObject*)original, (J9IndexableObject*)copy), objectMapFunction, objectMapData;
 #else /* defined(J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER) */
-		bool copyLockword = true;
-		
 		if (OBJECT_HEADER_SHAPE_POINTERS == J9CLASS_SHAPE(objectClass)) {
-			VM_ArrayCopyHelpers::referenceArrayCopy(currentThread, original, 0, copy, 0, size);
+			if (NULL != objectMapFunction) {
+				currentThread->javaVM->memoryManagerFunctions->j9gc_objaccess_cloneIndexableObject(currentThread, (J9IndexableObject*)original, (J9IndexableObject*)copy, objectMapFunction, objectMapData);
+			} else {
+				VM_ArrayCopyHelpers::referenceArrayCopy(currentThread, original, 0, copy, 0, size);
+			}
 		} else {
 			VM_ArrayCopyHelpers::primitiveArrayCopy(currentThread, original, 0, copy, 0, size, (((J9ROMArrayClass*)objectClass->romClass)->arrayShape & 0x0000FFFF));
 		}
-		if (copyLockword) {
-			/* zero lockword, if present */
-			j9objectmonitor_t *lockwordAddress = getLockwordAddress(currentThread, copy);
-			if (NULL != lockwordAddress) {
-				J9_STORE_LOCKWORD(currentThread, lockwordAddress, 0);
-			}
-		}
```

Comment:
this is unrelated simplification? ideally, should be a separate PR

Reply:
It's not required as part of this change, but I didn't want to propagate the initializeLockword variable through to the out of line helper. It can be done separately if you like, leaving a small inconsistency here (which doesn't really matter).